sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala

```diff
@@ -58,60 +61,127 @@ case class Statistics(
+ * @param avgLen average length of the values. For fixed-length types, this should be a constant.
+ * @param maxLen maximum length of the values. For fixed-length types, this should be a constant.
  */
-case class ColumnStat(statRow: InternalRow) {
+// TODO: decide if we want to use bigint to represent ndv and numNulls.
+case class ColumnStat(
+    ndv: Long,
+    min: Any,
+    max: Any,
+    numNulls: Long,
+    avgLen: Long,
+    maxLen: Long) {
 
-  def forNumeric[T <: AtomicType](dataType: T): NumericColumnStat[T] = {
-    NumericColumnStat(statRow, dataType)
-  }
-  def forString: StringColumnStat = StringColumnStat(statRow)
-  def forBinary: BinaryColumnStat = BinaryColumnStat(statRow)
-  def forBoolean: BooleanColumnStat = BooleanColumnStat(statRow)
+  /**
+   * Returns a map from string to string that can be used to serialize the column stats.
+   * The key is the name of the field (e.g. "ndv" or "min"), and the value is the string
+   * representation for the value. The deserialization side is defined in [[ColumnStat.fromMap]].
+   *
+   * As part of the protocol, the returned map always contains a key called "version".
+   */
+  def toMap: Map[String, String] = Map(
+    "version" -> "1",
+    "ndv" -> ndv.toString,
+    "min" -> min.toString,
+    "max" -> max.toString,
+    "numNulls" -> numNulls.toString,
+    "avgLen" -> avgLen.toString,
+    "maxLen" -> maxLen.toString
+  )
+}
+
+
+object ColumnStat extends Logging {
 
-  override def toString: String = {
-    // use Base64 for encoding
-    Base64.encodeBase64String(statRow.asInstanceOf[UnsafeRow].getBytes)
+  /** Returns true iff the we support gathering column statistics on column of the given type. */
+  def supportsType(dataType: DataType): Boolean = dataType match {
+    case _: NumericType | TimestampType | DateType | BooleanType => true
+    case StringType | BinaryType => true
+    case _ => false
   }
-}
 
-object ColumnStat {
-  def apply(numFields: Int, str: String): ColumnStat = {
-    // use Base64 for decoding
-    val bytes = Base64.decodeBase64(str)
-    val unsafeRow = new UnsafeRow(numFields)
-    unsafeRow.pointTo(bytes, bytes.length)
-    ColumnStat(unsafeRow)
+  /**
+   * Creates a [[ColumnStat]] object from the given map. This is used to deserialize column stats
+   * from some external storage. The serialization side is defined in [[ColumnStat.toMap]].
+   */
+  def fromMap(dataType: DataType, map: Map[String, String]): Option[ColumnStat] = {
+    val str2val: (String => Any) = dataType match {
+      case _: IntegralType => _.toLong
+      case _: DecimalType => Decimal(_)
+      case DoubleType | FloatType => _.toDouble
+      case BooleanType => _.toBoolean
+      case _ => identity
+    }
+
+    try {
+      Some(ColumnStat(
+        ndv = map("ndv").toLong,
+        min = str2val(map.get("min").orNull),
+        max = str2val(map.get("max").orNull),
+        numNulls = map("numNulls").toLong,
+        avgLen = map.getOrElse("avgLen", "1").toLong,
```

Comment:
For avgLen I think we should either have them be correct for fixed length types, or as cloud-fan suggests have it be an option.